### ISSUE TITLE
Not OK: perl 5.00563 on MSWin32-x86 4.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#927]
Migrated from [rt.perl.org#1870](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1870) (status was 'resolved')

Searchable as RT1870$

**Comment by p5pRT** [comment]
### From Paul.Moore@uk.origin-it.com



The failure is actually 5\.005\_63 \- but I can't run perlbug against that
for obvious reasons\.\.\.

Building with most of the same configuration as below for 5\.005\_03\,
basically Win32\, NT4 SP5\, MSVC 6 SP 3\.

Make options&#8203;:

USE\_ITHREADS

none of
USE\_THREADS
USE\_MULTI
USE\_OBJECT
USE\_IMP\_SYS

The build failed in compiling sv\.c\, with the attached errors\.

It looks like USE\_ITHREADS requires one of the other options \(USE\_MULTI\)\,
but doesn't enforce it\.

The line in error is
&nbsp;    ret = \(void\*\)\(\(\(char\*\)aTHXo\) \+ \(\(\(char\*\)v\) \- \(char\*\)proto\_perl\)\);

Presumably\, under this combination\, aTHXo isn't defined\, so the result is
a syntax error\.

A search of the sources indicates aTHXo is only used in a couple of
places\, and in the rest it seems "correct" \- is this just a typo which
should be some other aTHX macro?

Paul Moore\.

Microsoft \(R\) Program Maintenance Utility   Version 6\.00\.8168\.0
Copyright \(C\) Microsoft Corp 1988\-1998\. All rights reserved\.

&nbsp;    cl \-c \-I\.\. \-nologo \-Gf \-W3 \-I\.\.\\lib\\CORE \-I\.\\include \-I\. \-I\.\.
\-DWIN32 \-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT \-DPERLDLL \-DPERL\_CORE   \-O1
\-MD \-DNDEBUG  \-DUSE\_ITHREADS \-Fo\.\.\\sv\.obj \.\.\\sv\.c
sv\.c
\.\.\\sv\.c\(6304\) : error C2059&#8203;: syntax error : '\)'
\.\.\\sv\.c\(6311\) : error C2143&#8203;: syntax error : missing '\{' before '\*'
\.\.\\sv\.c\(6313\) : warning C4142&#8203;: benign redefinition of type
\.\.\\sv\.c\(6313\) : error C2371&#8203;: 'Perl\_ss\_dup' : redefinition; different basic
types
&nbsp;        \.\.\\proto\.h\(779\) : see declaration of 'Perl\_ss\_dup'
\.\.\\sv\.c\(6314\) : error C2039&#8203;: 'Tsavestack' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6315\) : error C2039&#8203;: 'Tsavestack\_ix' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6316\) : error C2039&#8203;: 'Tsavestack\_max' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6534\) : warning C4133&#8203;: 'return' : incompatible types \- from 'union
any \*' to 'int \*'
\.\.\\sv\.c\(6645\) : error C2039&#8203;: 'Idebug' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6654\) : error C2039&#8203;: 'Isv\_undef' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6654\) : error C2198&#8203;: 'Perl\_ptr\_table\_store' : too few actual
parameters
\.\.\\sv\.c\(6667\) : error C2039&#8203;: 'Isv\_no' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6667\) : error C2198&#8203;: 'Perl\_ptr\_table\_store' : too few actual
parameters
\.\.\\sv\.c\(6680\) : error C2039&#8203;: 'Isv\_yes' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6680\) : error C2198&#8203;: 'Perl\_ptr\_table\_store' : too few actual
parameters
\.\.\\sv\.c\(6686\) : error C2039&#8203;: 'Istrtab' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6686\) : error C2198&#8203;: 'Perl\_ptr\_table\_store' : too few actual
parameters
\.\.\\sv\.c\(6688\) : error C2039&#8203;: 'Icompiling' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6691\) : error C2039&#8203;: 'Icompiling' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6691\) : error C2198&#8203;: 'Perl\_ptr\_table\_store' : too few actual
parameters
\.\.\\sv\.c\(6694\) : error C2039&#8203;: 'Tcurcop' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6694\) : error C2198&#8203;: 'Perl\_any\_dup' : too few actual parameters
\.\.\\sv\.c\(6697\) : error C2039&#8203;: 'Iorigargc' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6702\) : error C2039&#8203;: 'Iorigargv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6702\) : error C2039&#8203;: 'Iorigargv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6702\) : error C2198&#8203;: 'Perl\_savepv' : too few actual parameters
\.\.\\sv\.c\(6704\) : error C2039&#8203;: 'Ienvgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6704\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6705\) : error C2039&#8203;: 'Iincgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6705\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6706\) : error C2039&#8203;: 'Ihintgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6706\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6707\) : error C2039&#8203;: 'Iorigfilename' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6707\) : error C2039&#8203;: 'Iorigfilename' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6707\) : error C2198&#8203;: 'Perl\_savepv' : too few actual parameters
\.\.\\sv\.c\(6708\) : error C2039&#8203;: 'Idiehook' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6708\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6709\) : error C2039&#8203;: 'Iwarnhook' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6709\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6712\) : error C2039&#8203;: 'Iminus\_c' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6713\) : error C2039&#8203;: 'Ipatchlevel' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6713\) : warning C4022&#8203;: 'memcpy' : pointer mismatch for actual
parameter 2
\.\.\\sv\.c\(6713\) : error C2198&#8203;: 'memcpy' : too few actual parameters
\.\.\\sv\.c\(6714\) : error C2039&#8203;: 'Ilocalpatches' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6715\) : error C2039&#8203;: 'Isplitstr' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6716\) : error C2039&#8203;: 'Ipreprocess' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6717\) : error C2039&#8203;: 'Iminus\_n' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6718\) : error C2039&#8203;: 'Iminus\_p' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6719\) : error C2039&#8203;: 'Iminus\_l' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6720\) : error C2039&#8203;: 'Iminus\_a' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6721\) : error C2039&#8203;: 'Iminus\_F' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6722\) : error C2039&#8203;: 'Idoswitches' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6723\) : error C2039&#8203;: 'Idowarn' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6724\) : error C2039&#8203;: 'Idoextract' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6725\) : error C2039&#8203;: 'Isawampersand' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6726\) : error C2039&#8203;: 'Iunsafe' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6727\) : error C2039&#8203;: 'Iinplace' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6727\) : error C2039&#8203;: 'Iinplace' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6727\) : error C2198&#8203;: 'Perl\_savepv' : too few actual parameters
\.\.\\sv\.c\(6728\) : error C2039&#8203;: 'Ie\_script' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6728\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6729\) : error C2039&#8203;: 'Iperldb' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6730\) : error C2039&#8203;: 'Iperl\_destruct\_level' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6734\) : error C2039&#8203;: 'Ibasetime' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6735\) : error C2039&#8203;: 'Iformfeed' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6735\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6737\) : error C2039&#8203;: 'Imaxsysfd' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6738\) : error C2039&#8203;: 'Imultiline' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6739\) : error C2039&#8203;: 'Istatusvalue' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6745\) : error C2039&#8203;: 'Istdingv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6745\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6746\) : error C2039&#8203;: 'Istderrgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6746\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6747\) : error C2039&#8203;: 'Idefgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6747\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6748\) : error C2039&#8203;: 'Iargvgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6748\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6749\) : error C2039&#8203;: 'Iargvoutgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6749\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6750\) : error C2039&#8203;: 'Iargvout\_stack' : is not a member of
'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6750\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6753\) : error C2039&#8203;: 'Ireplgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6753\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6756\) : error C2039&#8203;: 'Ierrgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6756\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6759\) : error C2039&#8203;: 'IDBgv' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6759\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6760\) : error C2039&#8203;: 'IDBline' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6760\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6761\) : error C2039&#8203;: 'IDBsub' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6761\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6762\) : error C2039&#8203;: 'IDBsingle' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6762\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6763\) : error C2039&#8203;: 'IDBtrace' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6763\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6764\) : error C2039&#8203;: 'IDBsignal' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6764\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6765\) : error C2039&#8203;: 'Ilineary' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6765\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6766\) : error C2039&#8203;: 'Idbargs' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6766\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6769\) : error C2039&#8203;: 'Tdefstash' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6769\) : error C2198&#8203;: 'Perl\_sv\_dup' : too few actual parameters
\.\.\\sv\.c\(6770\) : error C2039&#8203;: 'Tcurstash' : is not a member of 'interpreter'
&nbsp;        \.\.\\perl\.h\(2522\) : see declaration of 'interpreter'
\.\.\\sv\.c\(6770\) : fatal error C1003&#8203;: error count exceeds 100; stopping
compilation
NMAKE : fatal error U1077&#8203;: 'cl' : return code '0x2'
Stop\.

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00503:

Summary of my perl5 (5.0 patchlevel 5 subversion 03) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86
    uname=''
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cl.exe', optimize='-O2 -MD -DNDEBUG', gccversion=
    cppflags='-DWIN32'
    ccflags ='-O2 -MD -DNDEBUG -DWIN32 -D_CONSOLE -DNO_STRICT
-DHAVE_DES_FCRYPT '
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release -machine:x86'
    libpth=C:\VisualStudio\VC98\lib
    libs=R:\Libraries\Lib\MD\libdes.lib oldnames.lib kernel32.lib user32.lib
gdi32.lib  winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib PerlCRT.lib
    libc=PerlCRT.lib, so=dll, useshrplib=yes, libperl=perl.lib
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release
-machine:x86'

Locally applied patches:
    


@INC for perl 5.00503:
    C:\Applications\Perl\lib/MSWin32-x86
    C:\Applications\Perl\lib
    C:\Applications\Perl\site\lib/MSWin32-x86
    C:\Applications\Perl\site\lib
    .


Environment for perl 5.00503:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=C:\WINNT\system32;C:\WINNT;C:\UTILS;C:\Applications\Perl\Bin;C:\Applica
tions\Perl\Bin\MSWin32-x86;C:\VisualStudio\Common\Tools\WinNT;C:\VisualStudi
o\Common\MSDev98\Bin;C:\VisualStudio\Common\Tools;C:\VisualStudio\VC98\bin;C
:\Program Files\Mts;C:\MSSQL7\BINN;C:\DMINT40\Win32\Bin;C:\Program
Files\CWNT;C:\NTRESKIT;C:\Oracle\Ora81\bin;C:\Program
Files\Oracle\jre\1.1.7\bin;C:\APPLIC~1\Tcl\bin;C:\Applications\Mingw32\bin
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Further information\. Perl builds with USE\_MULTI\, but it fails tests\.
Basically\, Perl still thinks that fork\(\) is unimplemented\. Is this expected?

Paul\.

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Moore\, Paul 
Sent&#8203;: 09 December 1999 16&#8203;:04
To&#8203;: perl5\-porters@&#8203;perl\.org
Subject&#8203;: \[ID 19991209\.004\] Not OK&#8203;: perl 5\.00563 on MSWin32\-x86 4\.0


This is a build failure report for perl from paul\.moore@&#8203;uk\.origin\-it\.com\,
generated with the help of perlbug 1\.26 running under perl 5\.00503\.




---
\[Please enter your report here\]

The failure is actually 5\.005\_63 \- but I can't run perlbug against that
for obvious reasons\.\.\.

Building with most of the same configuration as below for 5\.005\_03\,
basically Win32\, NT4 SP5\, MSVC 6 SP 3\.

Make options&#8203;:

USE\_ITHREADS

none of
USE\_THREADS
USE\_MULTI
USE\_OBJECT
USE\_IMP\_SYS

The build failed in compiling sv\.c\, with the attached errors\.

It looks like USE\_ITHREADS requires one of the other options \(USE\_MULTI\)\,
but doesn't enforce it\.

The line in error is
&nbsp;    ret = \(void\*\)\(\(\(char\*\)aTHXo\) \+ \(\(\(char\*\)v\) \- \(char\*\)proto\_perl\)\);

Presumably\, under this combination\, aTHXo isn't defined\, so the result is
a syntax error\.

A search of the sources indicates aTHXo is only used in a couple of
places\, and in the rest it seems "correct" \- is this just a typo which
should be some other aTHX macro?

Paul Moore\.

\[\[ Rest of original deleted \]\]

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 09 Dec 1999 16&#8203;:04&#8203;:01 GMT\, "Moore\, Paul" wrote&#8203;:
>The failure is actually 5\.005\_63 \- but I can't run perlbug against that
>for obvious reasons\.\.\.
>
>Building with most of the same configuration as below for 5\.005\_03\,
>basically Win32\, NT4 SP5\, MSVC 6 SP 3\.
>
>Make options&#8203;:
>
>USE\_ITHREADS
>
>none of
>USE\_THREADS
>USE\_MULTI
>USE\_OBJECT
>USE\_IMP\_SYS
>
>The build failed in compiling sv\.c\, with the attached errors\.
>
>It looks like USE\_ITHREADS requires one of the other options \(USE\_MULTI\)\,
>but doesn't enforce it\.

I guess that's why there's the following disclaimer&#8203;:

&nbsp;    \# Beginnings of interpreter cloning/threads&#8203;: still rather rough\, fails
&nbsp;    \# tests\.  This should be enabled to get the fork\(\) emulation\.  Do not
&nbsp;    \# enable unless you know what you're doing\!

It shouldn't fail any tests though\, unless you enable TOP\_CLONE\,
which is designed to fail some tests\. :\-\)

USE\_ITHREADS is meant to be used with either USE\_OBJECT or USE\_MULTI\.
It could be auto\-enabled in future when you ask for USE\_THREADS with
either USE\_OBJECT or USE\_MULTI\.

>A search of the sources indicates aTHXo is only used in a couple of
>places\, and in the rest it seems "correct" \- is this just a typo which
>should be some other aTHX macro?

Nope\, it's not a typo\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 09 Dec 1999 16&#8203;:34&#8203;:03 GMT\, "Moore\, Paul" wrote&#8203;:
>Further information\. Perl builds with USE\_MULTI\, but it fails tests\.

Which ones exactly?  If it's posix\.t\#2\, see README\.win32\.

>Basically\, Perl still thinks that fork\(\) is unimplemented\. Is this expected?

Not if you enable USE\_MULTI/USE\_OBJECT\, USE\_ITHREADS\, and USE\_IMP\_SYS\.
See pp\.c/pp\_fork\(\)\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

From&#8203;: Gurusamy Sarathy \[mailto&#8203;:gsar@&#8203;ActiveState\.com\]
> 
> On Thu\, 09 Dec 1999 16&#8203;:34&#8203;:03 GMT\, "Moore\, Paul" wrote&#8203;:
> >Further information\. Perl builds with USE\_MULTI\, but it fails tests\.
> 
> Which ones exactly?  If it's posix\.t\#2\, see README\.win32\.
>

Nope\. It was the fork\.t tests \- failing with "fork is unimplemented"
messages\. I don't  have the exact results now\, as I worked out what I'd done
wrong \(see below\)\.\.\.

> >Basically\, Perl still thinks that fork\(\) is unimplemented\. 
> Is this expected?
> 
> Not if you enable USE\_MULTI/USE\_OBJECT\, USE\_ITHREADS\, and USE\_IMP\_SYS\.
> See pp\.c/pp\_fork\(\)\.

Yes\, I got to this combination in the end\. Now passes all tests\. Sorry for
the false alarm\.

BTW\, I noticed that t/io/pipe\.t \(and others\) still skips testing as
$Config\{d\_fork\} is undef\. Would it be reasonable to define this at some
point\, so that modules which test for fork like this will work in the
presence of "pseudo\-fork"? OTOH\, just changing config\.vc doesn't work\, as
this triggets a rebuild with HAVE\_FORK defined\, which is \*not\* right\.\.\.

Anyway\, at present t/io/pipe\.t fails anyway\, as it looks like the open\("|\-"\)
stuff is not implemented \(yet\)\.

On a related note\, what is a good way of testing if fork\(\) works? eval \{
fork \} seems inappropriate\, as it will actually fork if the test works\.
Using %Config\, assuming it is changed\, is quite a large overhead\. Testing
$^O is error\-prone and \(conceptually\) wrong\. Maybe the common approach of
just forking and letting people withyout fork diagnose the errors is
actually a realistic approach :\-\)

Cheers\,
Paul\.
